Wrappers/Python/cil/processors/MaskGenerator.py

```diff
@@ -0,0 +1,334 @@
+      - mean              test element-wise if 
+                          abs(A - mean(A)) < threshold_factor * std(A).
+                          Default value of threshold_factor is 3. If no axis is specified, 
+                          then operates over flattened array. Alternatively operates along axis specified 
+                          as dimension_label.
+      - median            test element-wise if 
+                          abs(A - median(A)) < threshold_factor * scaled MAD(A),
+                          scaled MAD is defined as c*median(abs(A-median(A))) where c=-1/(sqrt(2)*erfcinv(3/2))
+                          Default value of threshold_factor is 3. If no axis is specified, 
+                          then operates over flattened array. Alternatively operates along axis specified 
+                          as dimension_label.
+      - movmean           the same as mean but uses rolling mean with a specified window,
+                          default window value is 5
+      - movmedian         the same as mean but uses rolling median with a specified window,
+                          default window value is 5
+
+      '''
+
+    def __init__(self,
+                 mode = 'special_values',
+                 threshold_value = (None, None),
+                 quantiles = (None, None),
+                 threshold_factor = 3,
+                 window = 5,
+                 axis = None):
+
+        kwargs = {'mode': mode,
+                  'threshold_value': threshold_value,
+                  'threshold_factor': threshold_factor,
+                  'quantiles': quantiles,
+                  'window': window,
+                  'axis': axis}
+
+        super(MaskGenerator, self).__init__(**kwargs)
+    
+    def check_input(self, data):
+        
+        if not (issubclass(type(data), DataContainer)):
+            raise TypeError('Processor supports only following data types:\n' +
+                            ' - ImageData\n - AcquisitionData\n' +
+                            ' - DataContainer')
+        if self.mode not in ['special_values', 'nan', 'inf', 'threshold', 'quantile',\
+                             'mean', 'median', 'movmean', 'movmedian']:
+            raise Exception("Wrong mode. One of the following is expected:\n" + \
+                            "special_values, nan, inf, threshold, \n quantile, mean, median, movmean, movmedian")
+        
+        if self.axis is not None:
+                
+            if self.axis not in data.dimension_labels.values():
+                raise Exception("Wrong label is specified for axis. " + \
+                    "Expected dimensions are {}.".format(data.dimension_labels))
+
+        return True 
+
+    def process(self):
+
+        # get input DataContainer
+        data = self.get_input()
+        
+        # intialise mask with all ones
+        mask = numpy.ones(data.as_array().shape, dtype=numpy.bool)
```

Comment:
when you call `as_array()` it will convert the data container array to a numpy array every call, unless it already is a numpy array. For CIL objects currently they will always be numpy arrays so there won't be an issue, you'll just get a reference to the array as it is in memory, but if we ever change the underlying array or use this for non-CIL objects there will be an issue.

I would call `arr = data.as_array()` once at the start and use  `arr` throughout, then use `data.fill(arr)` at the end. If `arr` is just a reference to `data.array` fill will do nothing, so this is safe, if it is a converted and modified numpy array it'll overwrite the old data as intended.